### ISSUE TITLE
Race condition in worker.shutdown() in a cluster environment

### ISSUE THREAD

**Issue by laggyluke** [Automattic/kue#538]
So this took me some time and hair-pulling to debug.

Apparently when you have more than one worker process for the same queue, a race condition can prevent some of the workers to gracefully shut down, forcing the whole thing to wait for a timeout.

Some of the jobs I'm processing may take longer than an hour, which forces me to adjust a timeout accordingly, thus making this bug especially painful as it triggers even when the queue is empty.

Here's what's going on:

0. Let's say we start with a completely empty queue.
1. Process A does `queue.shutdown()` which calls `worker.shutdown()` for every worker.
2. Process B does the same.
3. Process A is almost done, hits [this line](https://github.com/LearnBoost/kue/blob/90d902b74f87edabb8c0e6055ebf9ee50dbf78b2/lib/queue/worker.js#L282), adding an item to `jobs` list.
4. Process B hits [this line](https://github.com/LearnBoost/kue/blob/90d902b74f87edabb8c0e6055ebf9ee50dbf78b2/lib/queue/worker.js#L246), then [this line](https://github.com/LearnBoost/kue/blob/90d902b74f87edabb8c0e6055ebf9ee50dbf78b2/lib/queue/worker.js#L252), setting `self.job = true`, then yielding to event loop while waiting for `ZPOP` I/O to complete (next line).
5. At this time Process B hits [this line](https://github.com/LearnBoost/kue/blob/90d902b74f87edabb8c0e6055ebf9ee50dbf78b2/lib/queue/worker.js#L294), so assumes there's a job running and subscribes to relevant events to be notified when the job is completed or failed.
6. In reality, once `ZPOP` I/O is done, `self.job` is reset to `null` on [this line](https://github.com/LearnBoost/kue/blob/90d902b74f87edabb8c0e6055ebf9ee50dbf78b2/lib/queue/worker.js#L255), but `worker.shutdown()` is not notified about it, so it waits for the whole timeout duration.

My (ugly) workaround was to add the following piece of code right between [these two lines](https://github.com/LearnBoost/kue/blob/90d902b74f87edabb8c0e6055ebf9ee50dbf78b2/lib/queue/worker.js#L290-L291):

```javascript
if (this.job === true) {
    return setTimeout(self.shutdown.bind(self, fn, timeout), 10);
}
```

But you probably can come up with a better fix.

**Comment by behrad** [comment]
thank you @laggyluke for the details, I'll check it...

**Comment by behrad** [comment]
Can you test/confirm that removing this line https://github.com/LearnBoost/kue/blob/90d902b74f87edabb8c0e6055ebf9ee50dbf78b2/lib/queue/worker.js#L282 prevents from this to happen !?

**Comment by laggyluke** [comment]
Yes, I can confirm that I can't reproduce this bug when that line is commented out.
Though it looks like this won't fix the underlying race condition, just make it more rare / difficult to trigger.

I thought about adding a 'job idle' event that will fire when zpop returns nothing, so that `shutdown` could subscribe to it as well, but it still looks a bit hackish... Anyway, I'm no expert, you should probably know better :)

**Comment by behrad** [comment]
@laggyluke  Could you please confirm this is not happening any more!?

**Comment by laggyluke** [comment]
I've tried installing kue from master, but unfortunately got the following error during workers shutdown:
```
TypeError: undefined is not a function
  at fn (/path/to/project/node_modules/kue/lib/kue.js:327:20)
  at _fn (/path/to/project/node_modules/kue/lib/queue/worker.js:301:9)
  at Worker.shutdown (/path/to/project/node_modules/kue/lib/queue/worker.js:309:16)
  at /path/to/project/node_modules/kue/lib/kue.js:346:22
  at Array.forEach (native)
  at Queue.shutdown (/path/to/project/node_modules/kue/lib/kue.js:344:20)
  [...]
```

I'll try to debug it more thoroughly tomorrow.

**Comment by behrad** [comment]
Have you read the history changes and latest docs?
we have some API breaking changes from 0.8.x to 0.9

**Comment by laggyluke** [comment]
Thanks, you're right - I've updated my code for the new API and it worked just fine.
I can confirm that this issue is now resolved.
Thanks again!

**Comment by behrad** [comment]
happy to hear that :)
